File: docs/leetcode/946/946.py
from typing import List

class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack = []
        start = 0
        for pop in popped:
            if start < len(pushed) and pushed[start] == pop:
                start += 1
            elif len(stack) > 0 and stack[-1] == pop:
                stack.pop()
            else:
                while (start < len(pushed) and pushed[start] != pop):
                    stack.append(pushed[start])
                    start += 1
                start += 1
        if len(stack) == 0:
            return True
        return False

File: leetcode/946/946.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

bool validateStackSequences(int* pushed, int pushedSize, int* popped, int poppedSize){
	int *stack = (int *)malloc(sizeof(int) * pushedSize);
	int top = 0;
	int start = 0;
    for (int i = 0; i < poppedSize; i++) {
        if (start < pushedSize && popped[i] == pushed[start]) {
            start++;
        } else if (top > 0 && stack[top - 1] == popped[i]) {
            top--;
        } else {
            while (start < pushedSize && popped[i] != pushed[start]) {
                stack[top++] = pushed[start++];
            }
            start++;
        }
    }
    if (top > 0) {
        return false;
    }

    return true;
}

int main()
{
    int pushed[] = {1,2,3,4,5};
    int poped[] = {4,5,3,2,1};
    bool ans = validateStackSequences(pushed, 5, poped, 5);
    return 0;
}